[Automatic pyramids levels fit]

floffy-f opened the pull request:
# Automatic fil for the pyramids level

We want our images to go to a minimal resolution of **100x100px** at the top of the pyramid.
Each level divides the resolution by two.

Let <img src="https://render.githubusercontent.com/render/math?math=h^i"> be the height of the ith image, and <img src="https://render.githubusercontent.com/render/math?math=w^i"> its width.

Then, let <img src="https://render.githubusercontent.com/render/math?math=h=\min_{i\le N}h^i"> and  <img src="https://render.githubusercontent.com/render/math?math=w=\min_{i\le N}w^i">.

Then the sequence of the steps of the pyramid for this minimal image is a geometric sequence, which writes : <img src="https://render.githubusercontent.com/render/math?math=h_n=\frac{h}{2^n}>100"> and <img src="https://render.githubusercontent.com/render/math?math=w_n=\frac{w}{2^n}>100">

As we want a single n integer value for both height and width for all images, we choose the following formula :
<img style="border : 5px solid black;" src="https://render.githubusercontent.com/render/math?math=n=\left\lfloor log_2(min(h,w)/100)\right\rfloor">.
The floor function helps us providing safety, avoiding to go further than 100 at the top stage of the pyramid.

For most good quality pictures, this corresponds to 4 levels. For minimal resolutions higher than `3200px`, we have 5 of them.
Lower than that, we usually have 2 or 3 levels.

The old `levels` field from `defaultParams` and `defaultParamsForm` has not been taken out of the code, but I chose to put them at `1` just in case. No important effects on the code imho.

floffy-f committed:
Automatic estimate of a good number of pyramid levels

mpizenberg commented on web-elm/src/Main.elm:
```diff
@@ -495,6 +496,15 @@ update msg model =
+                    model.params
+
+                optiLevels =
+                    getOptimumPyramids newLoaded
```
I'd suggest doing this the moment where we switch from the loading state to the state where all images are loaded so that this is done only once instead of each time one image is decoded.

mpizenberg commented on web-elm/src/Main.elm:
```diff
@@ -1898,7 +1974,9 @@ viewConfig ({ params, paramsForm, paramsInfo } as model) =
-                    , Element.text ("(default to " ++ String.fromInt defaultParams.levels ++ ")")
+
+                    -- , Element.text ("(default to " ++ String.fromInt defaultParams.levels ++ ")")
+                    -- /!\ Default number of pyramids removed in order to put a fitting number of pyramids
```
Indeed it's not fixed anymore. Though once images are loaded, we could compute the default number of pyramid levels based on the images size and store that info inside the model so that we can display it here.

floffy-f replied:
### Solution :
To solve this, I chose to store the default number of pyramids in the model directly.
It takes a little bit of memory, but I think it might be useful elsewhere (by that I mean not in the config tab) sometime.
This `Int` value is shown in the little text that I put back in the code.

To find its value, you have to look at `model.paramsForm.defaultLevels`, which is pretty convenient.

mpizenberg commented:
I've rebased your changes on the latest `main`. So now it should be easy to address some of the review comments I did without fear of more conflicts.

floffy-f committed:
[Auto-resolution for pyramids]
The optimum number of pyramids is only computed once, when all images
are loaded

floffy-f committed:
[Auto-resolution for pyramids]
The default number of pyramids is stored in the model directly.
It is then displayed in the little text of the config tab : "(defaults to
x)"

floffy-f committed:
[Formating]
elm-format applied for the branch auto-res
